TileStache/Goodies/VecTiles/transform.py

```diff
@@ -555,6 +556,111 @@ def _geom_dimensions(g):
+
+        return geoms
+
+    else:
+        return [shape]
+
+
+def _filter_geom_types(shape, keep_dim):
+    """
+    Return a geometry which consists of the geometries in
+    the input shape filtered so that only those of the
+    given dimension remain. Collapses any structure (e.g:
+    of geometry collections) down to a single or multi-
+    geometry.
+
+    >>> _filter_geom_types(GeometryCollection(), _POINT_DIMENSION).wkt
+    'GEOMETRYCOLLECTION EMPTY'
+    >>> _filter_geom_types(Point(0,0), _POINT_DIMENSION).wkt
+    'POINT (0 0)'
+    >>> _filter_geom_types(Point(0,0), _LINE_DIMENSION).wkt
+    'GEOMETRYCOLLECTION EMPTY'
+    >>> _filter_geom_types(Point(0,0), _POLYGON_DIMENSION).wkt
+    'GEOMETRYCOLLECTION EMPTY'
+    >>> _filter_geom_types(LineString([(0,0),(1,1)]), _LINE_DIMENSION).wkt
+    'LINESTRING (0 0, 1 1)'
+    >>> _filter_geom_types(Polygon([(0,0),(1,1),(1,0),(0,0)],[]), _POLYGON_DIMENSION).wkt
+    'POLYGON ((0 0, 1 1, 1 0, 0 0))'
+    >>> _filter_geom_types(shapely.wkt.loads('GEOMETRYCOLLECTION (POINT(0 0), LINESTRING(0 0, 1 1))'), _POINT_DIMENSION).wkt
+    'POINT (0 0)'
+    >>> _filter_geom_types(shapely.wkt.loads('GEOMETRYCOLLECTION (POINT(0 0), LINESTRING(0 0, 1 1))'), _LINE_DIMENSION).wkt
+    'LINESTRING (0 0, 1 1)'
+    >>> _filter_geom_types(shapely.wkt.loads('GEOMETRYCOLLECTION (POINT(0 0), LINESTRING(0 0, 1 1))'), _POLYGON_DIMENSION).wkt
+    'GEOMETRYCOLLECTION EMPTY'
+    >>> _filter_geom_types(shapely.wkt.loads('GEOMETRYCOLLECTION (POINT(0 0), GEOMETRYCOLLECTION (POINT(1 1), LINESTRING(0 0, 1 1)))'), _POINT_DIMENSION).wkt
+    'MULTIPOINT (0 0, 1 1)'
+    >>> _filter_geom_types(shapely.wkt.loads('GEOMETRYCOLLECTION (MULTIPOINT(-1 -1, 0 0), GEOMETRYCOLLECTION (POINT(1 1), POINT(2 2), GEOMETRYCOLLECTION (POINT(3 3))), LINESTRING(0 0, 1 1))'), _POINT_DIMENSION).wkt
+    'MULTIPOINT (-1 -1, 0 0, 1 1, 2 2, 3 3)'
+    >>> _filter_geom_types(shapely.wkt.loads('GEOMETRYCOLLECTION (LINESTRING(-1 -1, 0 0), GEOMETRYCOLLECTION (LINESTRING(1 1, 2 2), GEOMETRYCOLLECTION (POINT(3 3))), LINESTRING(0 0, 1 1))'), _LINE_DIMENSION).wkt
+    'MULTILINESTRING ((-1 -1, 0 0), (1 1, 2 2), (0 0, 1 1))'
+    >>> _filter_geom_types(shapely.wkt.loads('GEOMETRYCOLLECTION (POLYGON((-2 -2, -2 2, 2 2, 2 -2, -2 -2)), GEOMETRYCOLLECTION (LINESTRING(1 1, 2 2), GEOMETRYCOLLECTION (POLYGON((3 3, 0 0, 1 0, 3 3)))), LINESTRING(0 0, 1 1))'), _POLYGON_DIMENSION).wkt
+    'MULTIPOLYGON (((-2 -2, -2 2, 2 2, 2 -2, -2 -2)), ((3 3, 0 0, 1 0, 3 3)))'
+    """
+
+    # flatten the geometries, and keep the parts with the
+    # dimension that we want. each item in the parts list
+    # should be a single (non-multi) geometry.
+    parts = []
+    for g in _flatten_geoms(shape):
+        if _geom_dimensions(g) == keep_dim:
+            parts.append(g)
+
+    if len(parts) == 0:
+        # the only way we can signal an empty geometry, as
+        # MultiPoint([]) throws an exception.
+        return GeometryCollection()
```

Comment:
Nice catch - I didn't even try just `MultiPoint()`. I've extracted the constructors in 15ed909 - hopefully that looks better?